src/ring_theory/adjoin_root.lean

```diff
@@ -244,6 +256,19 @@ lemma alg_hom_subsingleton {S : Type*} [comm_ring S] [algebra R S] {r : R} :
 
 end adjoin_inv
 
+section prime
+
+instance [hf : fact (prime f)] : is_domain (adjoin_root f) :=
+(ideal.quotient.is_domain_iff_prime (span {f} : ideal R[X])).mpr $
+  (ideal.span_singleton_prime (fact_iff.mp hf).ne_zero).mpr (fact_iff.mp hf)
+
+instance [is_domain R] [hf : fact (prime f)] [hf' : fact (0 < f.degree)] :
```

Comment:
Why `fact (0 < f.degree)` and not a normal assumption? You need to write `haveI` below, so I don't see the point.

Reply:
Would the idiomatic way of writing it with a normal assumption be something like `theorem foo instance [is_domain R] (hf :(prime f)) (hf' : 0 < f.degree)  : no_zero_smul_divisors R (adjoin_root f)`, or do instances still work with normal assumptions?

Reply:
I mean that I am afraid this cannot be a useful instance, since Lean cannot find automatically `0 < f.degree`. Of course using `fact` as you did gives a theoretically usable instance, but in practice it will work very rarely, since Lean will not find `fact...`. My personal solution is just to leave it as a theorem with a comment that it is not an instance, so it must be used my hand with `haveI` (so no `fact`, in any case one `haveI` is necessary). For example as in [no_zero_smul_divisors.of_algebra_map_injective](https://leanprover-community.github.io/mathlib_docs/algebra/algebra/basic.html#no_zero_smul_divisors.of_algebra_map_injective).

On the other hand for fields it works without any assumption, this is worth to have as an instance (it can be already there).